/*
 * Copyright 2025 QingWan (qingwanmail@foxmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package qing.albatross.agent;

import android.app.Application;
import android.content.Context;

import qing.albatross.core.Albatross;
import qing.albatross.server.UnixRpcInstance;

public abstract class AlbatrossPlugin {

  protected String params;
  protected int flags;
  protected String libName;
  protected boolean enable;


  boolean hasFlags(int flags) {
    return (this.flags & flags) != 0;
  }

  public AlbatrossPlugin(String libName, String params, int flags) {
    this.params = params;
    this.flags = flags;
    this.libName = libName;
  }

  public String pluginName() {
    return this.getClass().getName();
  }

  public void loadLibrary(String libName) {
//    System.loadLibrary(libName); //The plugin only needs to remove the comments
    throw new RuntimeException("Plugins containing the so library must implement the loadLibrary method");
  }


  public boolean load(UnixRpcInstance agent) {
    try {
      if (libName != null)
        loadLibrary(libName);
      this.enable = true;
      return parseParams(params, flags);
    } catch (Throwable e) {
      this.enable = false;
      Albatross.log("plugin load err", e);
      return false;
    }
  }

  public boolean beforeApplicationCreateCall;

  public final boolean beforeApplicationCreateCall(Application application) {
    if (!beforeApplicationCreateCall) {
      beforeApplicationCreateCall = true;
      Albatross.log("call plugin " + pluginName() + " beforeApplicationCreate");
      try {
        beforeApplicationCreate(application);
      } catch (Exception e) {
        Albatross.log("call plugin " + pluginName() + " beforeApplicationCreate err", e);
      }
      return true;
    } else {
      Albatross.log("skip call plugin " + pluginName() + " beforeApplicationCreate");
      return false;
    }
  }

  public void beforeApplicationCreate(Application application) {
  }

  public boolean beforeMakeApplicationCall;

  public void beforeMakeApplication() {
  }

  public boolean beforeNewApplicationCall;

  public final synchronized boolean beforeNewApplicationCall(ClassLoader cl, String className, Context context) {
    if (!beforeNewApplicationCall) {
      beforeNewApplicationCall = true;
      Albatross.log("call plugin " + pluginName() + " beforeNewApplication");
      try {
        beforeNewApplication(cl, className, context);
      } catch (Exception e) {
        Albatross.log("call plugin " + pluginName() + " beforeNewApplication err", e);
      }
      return true;
    } else {
      Albatross.log("skip call plugin " + pluginName() + " beforeNewApplication");
    }
    return false;
  }

  public void beforeNewApplication(ClassLoader cl, String className, Context context) {

  }

  public boolean afterNewApplicationCall;

  public final boolean afterNewApplicationCall(Application application) {
    synchronized (this) {
      if (!afterNewApplicationCall) {
        afterNewApplicationCall = true;
      } else {
        Albatross.log("skip call plugin " + pluginName() + " afterNewApplication");
        return false;
      }
    }
    Albatross.log("call plugin " + pluginName() + " afterNewApplication");
    try {
      afterNewApplication(application);
    } catch (Exception e) {
      Albatross.log("call plugin " + pluginName() + " afterNewApplication err", e);
    }
    return true;
  }

  public void afterNewApplication(Application application) {

  }

  public boolean parseParams(String params, int flags) {
    return true;
  }

  public void onAttachSystem(Application application) {
  }

  public void onConfigChange(String config, int flags) {
    this.params = config;
    this.flags = flags;
    if (this.enable)
      parseParams(config, flags);
  }

  public void unload() {
    disable();
  }

  public void disable() {
    this.enable = false;
    Albatross.log("plugin " + this.pluginName() + " disable");
    parseParams(null, 0);
  }

  public void enable() {
    this.enable = true;
    parseParams(params, flags);
  }

  public boolean isEnable() {
    return this.enable;
  }


  public boolean afterApplicationCreateCall;

  public final synchronized boolean afterApplicationCreateCall(Application application) {
    if (!afterApplicationCreateCall) {
      afterApplicationCreateCall = true;
      Albatross.log("call plugin " + pluginName() + " afterApplicationCreate");
      try {
        afterApplicationCreate(application);
      } catch (Exception e) {
        Albatross.log("call plugin " + pluginName() + " afterApplicationCreate err", e);
      }
      return true;
    } else {
      Albatross.log("skip call plugin " + pluginName() + " afterApplicationCreate");
    }
    return false;
  }

  public void afterApplicationCreate(Application application) {
  }


  // native avoid inline
  public native void send(String msg);

  //native avoid inline
  public native void send(String msg, Throwable tr);

  public native void log(String msg);

  public int getFlags() {
    return flags;
  }
}
